codalab/server/authenticated_plugin.py

```diff
@@ -1,52 +1,87 @@
 
+class AuthenticationPlugin(ABC):
+    _NOT_AUTHENTICATED_ERROR = 'User is not authenticated'
+    _NOT_VERIFIED_ERROR = 'User is not verified'
+    _ACCESS_DENIED_ERROR = (
+        'User is not given access to this CodaLab instance. Please contact the admin for access.'
+    )
 
-class UserVerifiedPlugin(object):
+    api = 2
+
+    @abstractmethod
+    def apply(self, callback, route):
+        pass
+
+    def is_protected_mode(self):
+        return os.environ.get('CODALAB_PROTECTED_MODE') == 'True'
+
+    def user_is_authenticated(self):
+        return (
+            hasattr(request, 'user')
+            and request.user is not None
+            and request.user is not PUBLIC_USER
+        )
+
+    def check_user_verified(self):
+        if self.user_is_authenticated() and not request.user.is_verified:
+            if request.is_ajax:
+                abort(httplib.UNAUTHORIZED, AuthenticationPlugin._NOT_VERIFIED_ERROR)
+            else:
+                redirect(url('resend_key'))
+
+    def check_user_authenticated(self):
+        if not self.user_is_authenticated():
+            if request.is_ajax:
+                abort(httplib.UNAUTHORIZED, AuthenticationPlugin._NOT_AUTHENTICATED_ERROR)
+            else:
+                redirect_with_query('/account/login', {'next': request.url})
+
+    def check_has_access(self):
+        if not request.user.has_access:
+            if request.is_ajax:
+                abort(httplib.UNAUTHORIZED, AuthenticationPlugin._ACCESS_DENIED_ERROR)
+            else:
+                redirect_with_query('/account/login', {'next': request.url})
+
+
+class UserVerifiedPlugin(AuthenticationPlugin):
     """
     Fails the request if the user is authenticated but not verified.
 
-    The handling of AJAX requests is the same as above for AuthenticatedPlugin.
+    The handling of AJAX requests is the same as AuthenticatedPlugin.
     """
 
-    api = 2
-
     def apply(self, callback, route):
         def wrapper(*args, **kwargs):
-            if user_is_authenticated() and not request.user.is_verified:
-                if request.is_ajax:
-                    abort(httplib.UNAUTHORIZED, 'User is not verified')
-                else:
-                    redirect(url('resend_key'))
-
+            self.check_user_verified()
             return callback(*args, **kwargs)
 
         return wrapper
 
 
-class PublicUserPlugin(object):
+class PublicUserPlugin(AuthenticationPlugin):
     """
     Sets request.user to PUBLIC_USER if none set yet.
     """
 
-    api = 2
-
     def apply(self, callback, route):
         def wrapper(*args, **kwargs):
-            if not user_is_authenticated():
+            if not self.user_is_authenticated():
                 request.user = PUBLIC_USER
             return callback(*args, **kwargs)
 
         return wrapper
 
 
-class AuthenticatedPlugin(object):
+class AuthenticatedPlugin(AuthenticationPlugin):
```

Comment:
Can we remove `AuthenticatedPlugin` altogether? It seems that we should be fine to change the remaining endpoints where this plugin is used (such as /account/reset, etc.) to use `AuthenticatedProtectedPlugin` instead.

Reply:
I think we still need this to give users access to manage their accounts and view account info.